add slider

cpboyd opened the pull request:
I've been using this in my own project and felt I've gotten it to a point worth sharing.

I just updated the track markers portion to be more React-oriented. Previously, I just had the functions doing manual DOM manipulations, since I wasn't actually using discrete sliders in my project.

I still need to update the README, specifically in relation to the Props section.

cpboyd committed:
add slider

moog16 reviewed (changes requested):
@cpboyd thanks for adding this component! It would help greatly if you were to add a screenshot test for me to test/poke the code. Could you please add one? 

Also I will be asking for unit tests, but that can happen after the code review. How does that sound?

cpboyd commented:
@moog16 I just fixed some lint and typing issues.

I'm reading up on the screenshot tests at the moment.

cpboyd commented:
@moog16 I added a screenshot test, but I'm still not completely clear on everything involved.

I saw the list in **screenshot-test-urls.tsx**, but adding `'slider'` there didn't seem to add it to the test.  Is that in the Docker image itself?

Also, `switch` and `ripple` screenshot tests don't seem to currently be running either.
They're also not in the **screenshot-test-urls.tsx**.

Is there a way for me to view and validate the "golden" image? And where am I supposed to submit that, since [the instructions](https://github.com/material-components/material-components-web-react/blob/master/docs/screenshot-tests.md#step-2-run-a-docker-container) seem to indicate that I can't upload them via the Docker.

moog16 commented:
@cpboyd I have to actually create the golden. I'll get around to this today. The instructions is for you to test everything out. I need to update the contribute doc though (as this might be where the confusion is coming from)

moog16 commented on packages/slider/index.scss:
```diff
@@ -0,0 +1,23 @@
+// The MIT License
+//
+// Copyright (c) 2018 Google, Inc.
```
2019

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,455 @@
+      /**
+       * Registers an event handler on the root element for a given event.
+       */
+      registerInteractionHandler: (
```
We omit the register/deregister methods in mdc react and just use synthetic event handling. Please see https://github.com/material-components/material-components-web-react/blob/master/packages/floating-label/index.tsx#L94 as an example

cpboyd replied:
Do I need to re-implement [these handlers](https://github.com/material-components/material-components-web/blob/master/packages/mdc-slider/foundation.ts#L147-L153) in React?

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,455 @@
+    // need to use getElementsByClassName since tabIndicator could be
+    // a non-semantic element (span, i, etc.). This is a problem since refs to a non semantic elements
+    // return the instance of the component.
+    return this.sliderElement.current.getElementsByClassName(
```
Can you use ref's instead of this?

moog16 commented on packages/slider/package.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "@material/react-slider",
+  "version": "0.12.0",
```
0.0.0

moog16 reviewed (changes requested):
Getting warning in screenshots:
<img width="734" alt="Screen Shot 2019-05-29 at 10 10 15 AM" src="https://user-images.githubusercontent.com/579873/58576817-fe9c3280-81f9-11e9-947d-a3e624843cc2.png">

Looks like you need to remove this from the render method's otherProps

cpboyd commented:
@moog16 Commit aa0a057 basically replicates a substantial portion of the [mdc-slider/foundation.ts](https://github.com/material-components/material-components-web/blob/master/packages/mdc-slider/foundation.ts)

Further, I left the `registerBodyInteractionHandler` as I'm not sure there's a more React-like way to add handlers to the document body.

cpboyd commented:
I guess the rebase to master confused @googlebot.

moog16 commented:
Changed the base to rc0.14.0

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,677 @@
+
+  get adapter(): MDCSliderAdapter {
+    return {
+      /**
```
can you remove all the adapter comments? I appreciate the thoroughness however it makes the adapter 200 lines.

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,677 @@
+        propertyName: string,
+        value: string
+      ): void => {
+        this.setStylePropertyOn(
```
do we know that this won't be overridden on a rerender? This updates the style property on the DOM Node. I'm not sure if it will persist.

cpboyd replied:
I just based it on the [tab-indicator](https://github.com/material-components/material-components-web-react/blob/master/packages/tab-indicator/index.tsx#L118).

I'll look at refactoring this.

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,677 @@
+       * Sets a style property of the track element to the passed value
+       */
+      setTrackStyleProperty: (propertyName: string, value: string): void => {
+        this.setStylePropertyOn(this.trackElement.current, propertyName, value);
```
Here too...does this persist?

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,677 @@
+  /**
+   * Conditionally adds or removes a class based on shouldBePresent
+   */
+  private toggleClass(className: string, shouldBePresent: boolean) {
```
we shouldn't be calling adapter methods in our component.

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,677 @@
+      };
+    }
+
+    if (shouldFireInput) {
```
We shouldn't be using adapter methods, but instead relying on the foundation.

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,677 @@
+    // (See https://github.com/material-components/material-components-web/issues/1192)
+    const upHandler = () => {
+      this.handleUp();
+      this.adapter.deregisterBodyInteractionHandler(moveEventType, moveHandler);
```
This should be handled by the foundation. The general rule is that we need to implement all adapter methods (as you've done so above), and use all foundation methods. If we don't use the foundation methods, then that needs to be commented. 

Please refer to the vanilla web component to check this.

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,677 @@
+
+    // Prevent page from scrolling due to key presses that would normally scroll the page
+    evt.preventDefault();
+    this.adapter.addClass(cssClasses.FOCUS);
```
You should most likely be calling foundation.handleDown_. I know this is a private method, however you can refer to  https://github.com/material-components/material-components-web/issues/2813 in a comment when using a private event handler method from the foundation in this slider package.

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,677 @@
+          {discrete && (
+            <div className='mdc-slider__pin'>
+              <span className='mdc-slider__pin-value-marker'>
+                {markerValue}
```
The markerValue in the demo is a float/has many decimal places. I believe this is a bug. Please address

cpboyd replied:
I don't see this when rendering in Chrome, are you saying it has decimal places in the screenshot?

moog16 replied:
Yes, as you move the slider it displays decimal places.

cpboyd replied:
This is what I see in Chrome on macOS:
![image](https://user-images.githubusercontent.com/6323077/59272627-67d16c00-8c24-11e9-95f5-7f0628cdd941.png)

If it's still happening after I removed the `setValue` function in https://github.com/material-components/material-components-web-react/pull/868/commits/92d43bec99de0cdcda193cca2fefb78a3722d86d, then I can try to add code to truncate the value displayed by React.

I guess we'll have to account for the fact that the step itself can be a floating point number.

TroyTae commented:
#258

cpboyd commented:
@moog16 How do you suggest I handle the `onTransitionEnd` within 
[updateUIForCurrentValue_](https://github.com/material-components/material-components-web/blob/master/packages/mdc-slider/foundation.ts#L507)?

This gets called by `setValue_`, and thus most of the events.

cpboyd committed:
feat(slider): use private foundation functions

moog16 commented:
@cpboyd I see what you're saying about updateUIForCurrentValue. What if you do implement `[de]registerThumbContainerInteractionHandler`. You can set a flag to true in the register and set the flag to false in the deregister. Then in this.onThumbEnd, say:

```
onThumbEnd: () => {
  if (!this.flag) return;
  // ...
}
```

cpboyd commented:
@moog16 That's effectively what I tried to do in https://github.com/material-components/material-components-web-react/pull/868/commits/b333d05c08152eb700c4a6fd62d3e7e1881b783c

moog16 commented:
ah you're right...Sorry, it was a lot to sift through. But what you did makes sense.

moog16 commented:
Is this ready for review again?

cpboyd commented:
@moog16 I rebased with rc0.14.0.

Should I also squash the slider commits down?

moog16 commented:
no need! We'll squash merge this puppy when its ready. I'll take anotehr look. Thanks!

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+const convertDashToCamelCase = (propName: string) =>
+  propName.replace(/-(\w)/g, (_, v) => v.toUpperCase());
+
+type MouseLikeEvent = React.MouseEvent | React.PointerEvent | React.TouchEvent;
```
Let's call this InteractionEvent

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+interface SliderState {
+  classList: Set<string>;
+  markerValue?: number;
+  markerStyles: React.CSSProperties[];
```
why is this an array, but the others are not?

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+  thumbContainerStyleProperty: React.CSSProperties;
+}
+
+type SliderElementNames = 'trackStyleProperty' | 'thumbContainerStyleProperty';
```
SliderElementNames --> ElementStyleNames

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+    const {classList} = this.state;
+    const {className, discrete, displayMarkers} = this.props;
+    return classnames('mdc-slider', Array.from(classList), className, {
+      'mdc-slider--discrete': discrete,
```
can you use the constants? `cssClasses.DISCRETE`, as well as below

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+        }
+        return this.sliderElement.current.getBoundingClientRect();
+      },
+      getTabIndex: (): number => this.props.tabIndex || 0,
```
This defaults to 0. Change this to `number -> this.props.tabIndex`

cpboyd replied:
Had to force with `this.props.tabIndex!` since `HTMLDivElement` marks `tabIndex` as optional.

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+        handler: SpecificEventListener<EventType>
+      ): void => {
+        if (evtType === this.transitionendEvtName) {
+          this.onThumbEnd = handler as any;
```
can you change `any` to a specific event listener?

cpboyd replied:
I'll cast it to `() => void` since that's the type of the function.

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+          for (let i = 0; i < numMarkers; i++) {
+            markerStyles.push({});
+          }
+          return {markerStyles};
```
I don't think you need to return here.

moog16 replied:
Type in adapter is `appendTrackMarkers(numMarkers: number): void;`

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+          return {markerStyles};
+        });
+      },
+      removeTrackMarkers: (): void => {
```
remove this void...its unnecessary

cpboyd replied:
Should I remove the other void return specifications?

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+          return {markerStyles};
+        });
+      },
+      isRTL: (): boolean => this.props.isRtl || false,
```
|| false is unnecessary as you have it defined in defaultProps.

cpboyd replied:
Removed optional types for props per https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#support-for-defaultprops-in-jsx

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+    this.foundation.layout();
+    this.foundation.setupTrackMarker();
+
+    this.onDown = (evt) => (this.foundation as any).handleDown_(evt);
```
Why do you need to cast as `any`?

cpboyd replied:
To eliminate the Typescript complaints in VS Code about not knowing the `private` function.

It seems the `lint` check doesn't care, though, so I can remove it... I was just trying to resolve the majority of red underlines in VS Code.

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+
+type MouseLikeEvent = React.MouseEvent | React.PointerEvent | React.TouchEvent;
+
+export interface SliderProps extends React.HTMLProps<HTMLDivElement> {
```
Is it more appropriate for the HTMLDivElement to be HTMLInputElement?

cpboyd replied:
Hmm.  That's a good point.  The type of `step`, `min`, `max`, and `value` would need to change to `string`, but it might be good from a uniformity standpoint.

I think I'd considered `HTMLInputElement`, but went with `HTMLDivElement` because it doesn't actually use an `<input>` element, so we could potentially end up passing `HTMLInputElement`-specific props to the `<div>` via `otherProps` unless they're explicitly removed in the `props` destructuring.

moog16 commented on packages/slider/index.tsx:
```diff
@@ -0,0 +1,380 @@
+        aria-valuemax={max}
+        aria-valuenow={value}
+        data-step={step}
+        aria-label='Select Value'
```
we should allow for this to be overridable:

```
aria-label={this.props.ariaLabel}
```

set as a defaultProp to 'Select Value'

moog16 reviewed (changes requested):
Hey looking good! Two things I noticed:

- the onValueChange doesn't fire when I programatically change `value` on one of the screenshot tests. I think there isn't a way to achieve this with the current foundation. 
- also need unit tests. We aim for 95% coverage. If you run npm run test:unit, this will build a coverage directory. If you open the index.html file within there, it will show you the coverage.

cpboyd committed:
style(slider): remove void return specifiers

cpboyd committed:
refactor(slider): use cssClasses; rename types

cpboyd commented:
@moog16 I think I've addressed all the new comments.
I'll work on the unit tests.

moog16 commented:
Great! Let me know when its ready for review :)

cpboyd committed:
feat(slider): add basic unit tests

cpboyd commented:
@moog16 I rebased to rc0.15 and added some unit tests.

Note: I'm currently using `@material/slider` 2.3.0.  I'm not overly familiar with `material-components-web` versioning.  Some of the React components are still using the mdc v1 versions.  Should I downgrade (or optionally upgrade to v3)?
